File: reth/reth_chain_query/examples/entities/stablecoin_market_by_unit.rs
/// Stablecoin Market Analysis by Currency Unit
/// 
/// Demonstrates market share analysis grouped by currency unit (USD, EUR, JPY, etc.)
/// Each unit shows totals in its own denomination without cross-currency conversion.
/// This provides accurate representation of each currency's stablecoin ecosystem.

use reth_chain_query::{ChainQuery, Result};
use reth_chain_query::entities::stablecoins::StablecoinMarketAnalyzer;
use std::sync::Arc;

#[tokio::main]
async fn main() -> Result<()> {
    println!("💱 Stablecoin Market Analysis by Currency Unit");
    println!("{}", "=".repeat(70));
    
    // Initialize ChainQuery
    let reth_datadir = "/home/nima/.local/share/reth/mainnet";
    let chain_query = Arc::new(ChainQuery::new(reth_datadir)?);
    
    // Create market analyzer
    let analyzer = StablecoinMarketAnalyzer::new(chain_query.clone());
    
    // Get latest block
    let latest_block = chain_query.get_latest_block()?;
    println!("\n📊 Analyzing at block: {}", latest_block);
    println!("{}", "-".repeat(70));
    
    // Analyze market by currency unit
    let market_by_unit = analyzer.analyze_market_by_unit(Some(latest_block)).await?;
    
    // Display results for each currency unit
    println!("\n🌍 MARKET BY CURRENCY UNIT");
    println!("{}", "=".repeat(70));
    
    // Sort units by total supply for better display
    let mut units: Vec<_> = market_by_unit.units.iter().collect();
    units.sort_by(|a, b| {
        b.1.total_supply_in_unit
            .partial_cmp(&a.1.total_supply_in_unit)
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    
    for (unit_name, unit_data) in units {
        // Skip units with no supply
        if unit_data.total_supply_in_unit == 0.0 {
            continue;
        }
        
        println!("\n💵 {} Market", unit_name);
        println!("{}", "-".repeat(50));
        
        // Format total based on currency
        let total_formatted = match unit_name.as_str() {
            "US Dollar" => format!("${:.2}", unit_data.total_supply_in_unit),
            "Euro" => format!("€{:.2}", unit_data.total_supply_in_unit),
            "Japanese Yen" => format!("¥{:.0}", unit_data.total_supply_in_unit),
            "British Pound" => format!("£{:.2}", unit_data.total_supply_in_unit),
            "Singapore Dollar" => format!("S${:.2}", unit_data.total_supply_in_unit),
            "Australian Dollar" => format!("A${:.2}", unit_data.total_supply_in_unit),
            "Canadian Dollar" => format!("C${:.2}", unit_data.total_supply_in_unit),
            "Chinese Yuan" => format!("¥{:.2}", unit_data.total_supply_in_unit),
            "Indonesian Rupiah" => format!("Rp {:.0}", unit_data.total_supply_in_unit),
            _ => format!("{:.2} {}", unit_data.total_supply_in_unit, unit_name),
        };
        
        println!("Total Supply: {}", total_formatted);
        println!("Active Tokens: {}", unit_data.tokens.len());
        
        // Show top tokens in this unit
        println!("\nTop Stablecoins:");
        for (i, token) in unit_data.tokens.iter().take(5).enumerate() {
            let rank = i + 1;
            println!("  {}. {} ({})", 
                rank,
                token.info.symbol,
                token.info.name
            );
            
            // Format amount based on currency
            let amount_formatted = match unit_name.as_str() {
                "US Dollar" => format!("${:.2}", token.total_supply_formatted),
                "Euro" => format!("€{:.2}", token.total_supply_formatted),
                "Japanese Yen" => format!("¥{:.0}", token.total_supply_formatted),
                "British Pound" => format!("£{:.2}", token.total_supply_formatted),
                "Singapore Dollar" => format!("S${:.2}", token.total_supply_formatted),
                "Australian Dollar" => format!("A${:.2}", token.total_supply_formatted),
                "Canadian Dollar" => format!("C${:.2}", token.total_supply_formatted),
                "Chinese Yuan" => format!("¥{:.2}", token.total_supply_formatted),
                "Indonesian Rupiah" => format!("Rp {:.0}", token.total_supply_formatted),
                _ => format!("{:.2}", token.total_supply_formatted),
            };
            
            println!("     Supply: {}", amount_formatted);
            println!("     Market Share (within {}): {:.2}%", unit_name, token.market_share_percent);
        }
        
        // Show concentration metrics for this unit
        if unit_data.tokens.len() >= 3 {
            let top_3_share: f64 = unit_data.tokens
                .iter()
                .take(3)
                .map(|t| t.market_share_percent)
                .sum();
            
            println!("\nConcentration Metrics:");
            println!("  Top 3 Control: {:.2}% of {} market", top_3_share, unit_name);
            
            if top_3_share > 80.0 {
                println!("  ⚠️  Highly concentrated market");
            } else if top_3_share > 60.0 {
                println!("  ⚡ Moderately concentrated market");
            } else {
                println!("  ✅ Competitive market");
            }
        }
    }
    
    // Summary statistics
    println!("\n📈 OVERALL SUMMARY");
    println!("{}", "=".repeat(70));
    
    let active_units = units.iter()
        .filter(|(_, data)| data.total_supply_in_unit > 0.0)
        .count();
    
    let total_tokens: usize = units.iter()
        .map(|(_, data)| data.tokens.len())
        .sum();
    
    println!("Active Currency Units: {}", active_units);
    println!("Total Active Stablecoins: {}", total_tokens);
    
    // Show unit diversity
    println!("\nUnit Diversity:");
    for (unit_name, unit_data) in &units {
        if unit_data.total_supply_in_unit > 0.0 {
            let token_count = unit_data.tokens.len();
            let emoji = if token_count > 5 { "🔥" } 
                else if token_count > 2 { "📊" }
                else { "🌱" };
            
            println!("  {} {}: {} token(s)", emoji, unit_name, token_count);
        }
    }
    
    // Dominant unit analysis
    if let Some((dominant_unit, dominant_data)) = units.first() {
        println!("\n🏆 Dominant Currency Unit: {}", dominant_unit);
        println!("  Has {} active stablecoins", dominant_data.tokens.len());
        
        if let Some(top_token) = dominant_data.tokens.first() {
            println!("  Led by: {} ({:.2}% of {} market)", 
                top_token.info.symbol,
                top_token.market_share_percent,
                dominant_unit
            );
        }
    }
    
    println!("\n✅ Market analysis by currency unit complete!");
    
    println!("\n💡 Key Insights:");
    println!("- Each currency unit's market is analyzed independently");
    println!("- Market shares are calculated WITHIN each unit (not across units)");
    println!("- No cross-currency conversion is performed");
    println!("- This provides accurate representation of each currency's ecosystem");
    
    Ok(())
}

File: pyeth/eth_data/scripts/data/entities/stablecoins/stablecoin_market_by_unit.py
#!/usr/bin/env python3
"""
Stablecoin Market Analysis by Currency Unit using PyReth

Demonstrates how to analyze stablecoin markets grouped by currency unit (USD, EUR, JPY, etc.)
Each unit shows totals in its own denomination without cross-currency conversion.
"""

from pyreth import PyReth
from typing import Dict, List
import sys


def format_currency_amount(amount: float, unit_name: str) -> str:
    """Format amount based on currency unit"""
    formatters = {
        "US Dollar": lambda x: f"${x:,.2f}",
        "Euro": lambda x: f"€{x:,.2f}",
        "Japanese Yen": lambda x: f"¥{x:,.0f}",
        "British Pound": lambda x: f"£{x:,.2f}",
        "Singapore Dollar": lambda x: f"S${x:,.2f}",
        "Australian Dollar": lambda x: f"A${x:,.2f}",
        "Canadian Dollar": lambda x: f"C${x:,.2f}",
        "Chinese Yuan": lambda x: f"¥{x:,.2f}",
        "Indonesian Rupiah": lambda x: f"Rp {x:,.0f}",
    }
    
    formatter = formatters.get(unit_name, lambda x: f"{x:,.2f} {unit_name}")
    return formatter(amount)


def analyze_stablecoin_market_by_unit():
    """Analyze stablecoin market by currency unit"""
    
    print("💱 Stablecoin Market Analysis by Currency Unit (PyReth)")
    print("=" * 70)
    
    # Initialize PyReth with singleton pattern
    reth = PyReth()
    chain_query = reth.chain_query()
    
    # Get latest block
    latest_block = chain_query.get_latest_block()
    print(f"\n📊 Analyzing at block: {latest_block}")
    print("-" * 70)
    
    # Get entity statistics first
    entity_stats = chain_query.get_entity_stats()
    print(f"\n📈 Tracking {entity_stats['stablecoin_count']} stablecoins")
    
    # Analyze market by currency unit
    market_by_unit = chain_query.analyze_stablecoin_market_by_unit()
    
    print("\n🌍 MARKET BY CURRENCY UNIT")
    print("=" * 70)
    
    # Sort units by total supply
    units = market_by_unit['units']
    sorted_units = sorted(
        units.items(),
        key=lambda x: x[1]['total_supply_in_unit'],
        reverse=True
    )
    
    # Display each currency unit
    for unit_name, unit_data in sorted_units:
        # Skip units with no supply
        if unit_data['total_supply_in_unit'] == 0:
            continue
        
        print(f"\n💵 {unit_name} Market")
        print("-" * 50)
        
        # Format total based on currency
        total_formatted = format_currency_amount(
            unit_data['total_supply_in_unit'],
            unit_name
        )
        
        print(f"Total Supply: {total_formatted}")
        print(f"Active Tokens: {len(unit_data['tokens'])}")
        
        # Show top tokens in this unit
        if unit_data['tokens']:
            print("\nTop Stablecoins:")
            for i, token in enumerate(unit_data['tokens'][:5], 1):
                print(f"  {i}. {token['symbol']}")
                
                # Format amount based on currency
                amount_formatted = format_currency_amount(
                    token['total_supply'],
                    unit_name
                )
                
                print(f"     Supply: {amount_formatted}")
                print(f"     Market Share (within {unit_name}): {token['market_share_percent']:.2f}%")
        
        # Show concentration metrics for this unit
        if len(unit_data['tokens']) >= 3:
            top_3_share = sum(
                token['market_share_percent'] 
                for token in unit_data['tokens'][:3]
            )
            
            print("\nConcentration Metrics:")
            print(f"  Top 3 Control: {top_3_share:.2f}% of {unit_name} market")
            
            if top_3_share > 80:
                print("  ⚠️  Highly concentrated market")
            elif top_3_share > 60:
                print("  ⚡ Moderately concentrated market")
            else:
                print("  ✅ Competitive market")
    
    # Summary statistics
    print("\n📈 OVERALL SUMMARY")
    print("=" * 70)
    
    active_units = sum(1 for _, data in sorted_units if data['total_supply_in_unit'] > 0)
    total_tokens = sum(len(data['tokens']) for _, data in sorted_units)
    
    print(f"Active Currency Units: {active_units}")
    print(f"Total Active Stablecoins: {total_tokens}")
    
    # Show unit diversity
    print("\nUnit Diversity:")
    for unit_name, unit_data in sorted_units:
        if unit_data['total_supply_in_unit'] > 0:
            token_count = len(unit_data['tokens'])
            if token_count > 5:
                emoji = "🔥"
            elif token_count > 2:
                emoji = "📊"
            else:
                emoji = "🌱"
            
            print(f"  {emoji} {unit_name}: {token_count} token(s)")
    
    # Dominant unit analysis
    if sorted_units:
        dominant_unit, dominant_data = sorted_units[0]
        print(f"\n🏆 Dominant Currency Unit: {dominant_unit}")
        print(f"  Has {len(dominant_data['tokens'])} active stablecoins")
        
        if dominant_data['tokens']:
            top_token = dominant_data['tokens'][0]
            print(f"  Led by: {top_token['symbol']} ({top_token['market_share_percent']:.2f}% of {dominant_unit} market)")
    
    print("\n✅ Market analysis by currency unit complete!")
    
    print("\n💡 Key Insights:")
    print("- Each currency unit's market is analyzed independently")
    print("- Market shares are calculated WITHIN each unit (not across units)")
    print("- No cross-currency conversion is performed")
    print("- This provides accurate representation of each currency's ecosystem")
    
    # Also demonstrate overall market share analysis
    print("\n" + "=" * 70)
    print("📊 OVERALL MARKET SHARE ANALYSIS")
    print("=" * 70)
    
    market_share = chain_query.analyze_stablecoin_market_share()
    
    print(f"\nTotal Market Supply (USD equivalent): ${market_share['total_market_supply']:,.2f}")
    print(f"Top 3 Concentration: {market_share['top_3_concentration']:.2f}%")
    print(f"Top 5 Concentration: {market_share['top_5_concentration']:.2f}%")
    print(f"Herfindahl Index: {market_share['herfindahl_index']:.2f}")
    
    print("\nTop 5 Stablecoins Overall:")
    for i, coin in enumerate(market_share['stablecoins'][:5], 1):
        print(f"  {i}. {coin['symbol']} - {coin['unit']}")
        print(f"     Supply: {coin['total_supply']:,.2f}")
        print(f"     Market Share: {coin['market_share_percent']:.2f}%")


def test_entity_identification():
    """Test entity identification capabilities"""
    
    print("\n" + "=" * 70)
    print("🔍 ENTITY IDENTIFICATION TEST")
    print("=" * 70)
    
    reth = PyReth()
    chain_query = reth.chain_query()
    
    test_addresses = [
        ("0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48", "USDC"),
        ("0xdAC17F958D2ee523a2206206994597C13D831ec7", "USDT"),
        ("0x28C6c06298d514Db089934071355E5743bf21d60", "Binance"),
        ("0x0171F896002665C3ea3Ed0c55f21026cA0A734A0", "BlackRock ETF"),
    ]
    
    for address, expected in test_addresses:
        entity_type = chain_query.identify_entity(address)
        
        emoji = {
            "stablecoin": "🪙",
            "cex": "🏦",
            "etf": "📈",
            "unknown": "❓"
        }.get(entity_type, "❓")
        
        print(f"\n{emoji} {address[:10]}...")
        print(f"  Expected: {expected}")
        print(f"  Type: {entity_type}")
        
        # Get detailed info based on type
        if entity_type == "stablecoin":
            info = chain_query.get_stablecoin_info(address)
            if info:
                print(f"  Details: {info['symbol']} ({info['unit']}) - {info['decimals']} decimals")
        elif entity_type == "cex":
            info = chain_query.get_cex_info(address)
            if info:
                print(f"  Details: {info['name']} ({info['exchange']})")
        elif entity_type == "etf":
            info = chain_query.get_etf_info(address)
            if info:
                print(f"  Details: {info['name']} ({info['provider']})")


if __name__ == "__main__":
    try:
        analyze_stablecoin_market_by_unit()
        test_entity_identification()
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)